src/server/android/android.ts

```diff
@@ -0,0 +1,280 @@
+      }
+    }
+
+    debug('pw:android')('Connected to driver');
+    const transport = new Transport(socket, socket, socket, 'be');
+    transport.onmessage = message => {
+      const response = JSON.parse(message);
+      const { id, result, error } = response;
+      const callback = this._callbacks.get(id);
+      if (!callback)
+        return;
+      if (error)
+        callback.reject(new Error(error));
+      else
+        callback.fulfill(result);
+      this._callbacks.delete(id);
+    };
+
+    callback(transport);
+    return this._driverPromise;
+  }
+
+  async send(method: string, params: any): Promise<any> {
+    const driver = await this._driver();
+    const id = ++this._lastId;
+    const result = new Promise((fulfill, reject) => this._callbacks.set(id, { fulfill, reject }));
+    driver.send(JSON.stringify({ id, method, params }));
+    return result;
+  }
+
+  async close() {
+    const driver = await this._driver();
+    driver.close();
+    await this._backend.close();
+  }
+
+  async launchBrowser(packageName: string = 'com.android.chrome', options: types.BrowserContextOptions = {}): Promise<BrowserContext> {
+    debug('pw:android')('Force-stopping', packageName);
+    await this._backend.runCommand(`shell:am force-stop ${packageName}`);
+
+    const socketName = createGuid();
+    const commandLine = `_ --disable-fre --no-default-browser-check --no-first-run --remote-debugging-socket-name=${socketName}`;
+    debug('pw:android')('Starting', packageName, commandLine);
+    await this._backend.runCommand(`shell:echo "${commandLine}" > /data/local/tmp/chrome-command-line`);
+    await this._backend.runCommand(`shell:am start -n ${packageName}/com.google.android.apps.chrome.Main about:blank`);
+
+    debug('pw:android')('Polling for socket', socketName);
+    while (true) {
+      const net = await this._backend.runCommand(`shell:cat /proc/net/unix | grep ${socketName}$`);
+      if (net)
+        break;
+      await new Promise(f => setTimeout(f, 100));
+    }
+    debug('pw:android')('Got the socket, connecting');
+    const androidBrowser = new AndroidBrowser(this, packageName, socketName);
+    await androidBrowser._open();
+
+    const browserOptions: BrowserOptions = {
+      name: 'clank',
+      slowMo: 0,
+      persistent: { ...options, noDefaultViewport: true },
+      downloadsPath: undefined,
+      browserProcess: new ClankBrowserProcess(androidBrowser),
+      proxy: options.proxy,
+      protocolLogger: helper.debugProtocolLogger(),
+      browserLogsCollector: new RecentLogsCollector()
+    };
+    validateBrowserContextOptions(options, browserOptions);
+
+    const browser = await CRBrowser.connect(androidBrowser, browserOptions);
+    const controller = new ProgressController();
+    await controller.run(async progress => {
+      await browser._defaultContext!._loadDefaultContext(progress);
+    });
+    return browser._defaultContext!;
+  }
+}
+
+class AndroidBrowser extends EventEmitter {
```

Comment:
Shall we mark this as `implements Transport`? IIUC, this serves as a transport for protocol connection?